Modules/fcntlmodule.c

```diff
@@ -89,6 +100,33 @@ fcntl_fcntl_impl(PyObject *module, int fd, int code, PyObject *arg)
         }
     }
 
+#ifdef F_KINFO
+    if (code == F_KINFO) {
+        struct kinfo_file f = {.kf_structsize = KINFO_FILE_SIZE};
+        do {
+            Py_BEGIN_ALLOW_THREADS
+                ret = fcntl(fd, code, &f);
+            Py_END_ALLOW_THREADS
+        } while (ret == -1 && errno == EINTR && !(async_err = PyErr_CheckSignals()));
+        if (ret < 0) {
+            return !async_err ? PyErr_SetFromErrno(PyExc_OSError) : NULL;
+        }
+
+        PyObject *ret = PyDict_New();
+        if (ret) {
+            PyDict_SetItemString(ret, "status", PyLong_FromLong(f.kf_status));
+            PyDict_SetItemString(ret, "type", PyLong_FromLong(f.kf_type));
+            PyDict_SetItemString(ret, "offset", PyLong_FromLongLong(f.kf_offset));
+            PyDict_SetItemString(ret, "path", PyBytes_FromString(f.kf_path));
+            return ret;
```

Comment:
``fcntl.fcntl()`` returns either an integer or bytes. If ``F_KINFO`` needs a different return type, then it would be better to implement the feature as a new platform-specific function in the fcntl module, e.g. ``fcntl.kinfo(fd: int) -> dict``.